Change default values of AnomalyLikelihood constructor args to match the empirically-optimized values in numenta-apps/htmengine's AnomalyLikelihoodHelper

vitaly-krugl opened the pull request:
Changed default values of AnomalyLikelihood constructor args to match the empirically-optimized values in numenta-apps/htmengine's AnomalyLikelihoodHelper.

Added reestimationPeriod arg to AnomalyLikelihood constructor.

Added logic in AnomalyLikelihood.anomalyProbability() to force greedy refresh of the
distribution when anomaly probability > 0.99 to match the empirically-optimized logic
in numenta-apps/htmengine's AnomalyLikelihoodHelper.

Implemented additional unit tests for AnomalyLikelihood class.

Deleted the recently-added AnomalyLikelihood.forceModelStateRefresh() method.

Fixed pylint findings in anomaly_likelihood_test.py.

@scottpurdy: please review

Please make sure this change gets into the nupic release of v0.2.13, since it is needed by https://github.com/numenta/numenta-apps/pull/277.

CC @rhyolight @jcasner

scottpurdy commented on nupic/algorithms/anomaly_likelihood.py:
```diff
@@ -204,18 +201,22 @@ def anomalyProbability(self, value, anomalyScore, timestamp=None):
-      )
       likelihood = 1.0 - likelihoods[0]
 
+      # If anomaly score > 0.99 then we greedily update the statistics.
```
This says why it is ok but not why we do it in the first place. I believe that it is to mitigate the impact of not updating the distribution at every iteration. It basically says let us only periodically update it but if we have a very high anomaly likelihood then we need it to be accurate so override the periodic update and do an update now.

scottpurdy commented on nupic/algorithms/anomaly_likelihood.py:
```diff
@@ -84,18 +88,20 @@ def __init__(self,
-
-    Anomaly likelihood scores are reported at a flat 0.5 for claLearningPeriod +
-    estimationSamples iterations.
+    data points to maintain for periodic reestimation of the Gaussian. Note:
```
I find it much easier to include indentation for continued param descriptions but I don't think there is an explicit convention.

vitaly-krugl replied:
Me, too. I was following the convention of the existing arg descriptions. I'll go ahead and fix this up. In reference to yesterday's "clean code" discussion, I noticed that this file is not using the doxygen convention for args, etc.

scottpurdy commented on nupic/algorithms/anomaly_likelihood.py:
```diff
@@ -166,12 +168,7 @@ def _calcSkipRecords(numIngested, windowSize, learningPeriod):
-  def forceModelRefresh(self):
-    """Force refresh of the anomaly likelihood model's state"""
-    self._distribution = None
+    return min(numIngested, max(0, learningPeriod - numShiftedOut))
```
good catch

vitaly-krugl replied:
I caught this once I implemented unit tests for `_calcSkipRecords`. I think that the user of the value handled it gracefully, but it was potentially error-prone and unintuitive.

scottpurdy commented on nupic/algorithms/anomaly_likelihood.py:
```diff
@@ -106,11 +112,7 @@ def __init__(self,
-    # system is not very sensitive to this number as long as it is small
-    # relative to the total number of records processed.
-    self._reestimationPeriod = 100
+    self._reestimationPeriod = reestimationPeriod
```
do we need to update __setstate__ to handle old saved instances of this? Or is it not being saved by anyone yet? Probably better to be safe.

vitaly-krugl replied:
@scottpurdy: a couple of comments regarding `__setstate__`:

1. The same member variable `_reestimationPeriod` existed before; if someone saved it having the value of 100, then it would have been acceptable to them, and there is no need to fix it up upon de-serialization.
2. Also, `__setstate__` doesn't exist for this class, anyway. But I think that the above item is what really matters.

scottpurdy replied:
So old saved versions would still work fine? Sounds good.

vitaly-krugl replied:
I think that the old saved versions should work fine.

scottpurdy commented on tests/unit/nupic/algorithms/anomaly_likelihood_test.py:
```diff
@@ -610,16 +654,185 @@ def testEstimationWindowSize(self):
     l2 = an.AnomalyLikelihood(claLearningPeriod=2, estimationSamples=2)
     self.assertEqual(l, l2)
 
+    # use 5>2+2 probationary samples to create distribution estimate
```
This is obviously testing a specific corner case but you could make it easier on the reader by stating what the case is rather than assuming they understand what `5>2+2` is referring to

vitaly-krugl replied:
@scottpurdy: This was an existing comment that I moved to make pylint happy. I will figure out the meaning and improve the comment.

scottpurdy commented on tests/unit/nupic/algorithms/anomaly_likelihood_test.py:
```diff
@@ -610,16 +654,185 @@ def testEstimationWindowSize(self):
     self.assertEqual(len(l._historicalScores), 3)
 
 
+  def testAnomalyLikelihoodWindowSizeImpactOnDistributionEstimateScoreArgs(
```
These tests are in the class `AnomalyLikelihoodTest` so I think you can leave "AnomalyLikelihood" out of the names.

vitaly-krugl replied:
@scottpurdy: it's actually more confusing than that: the `AnomalyLikelihoodTest` class contains test methods for the low-level functions as well as for the high-level `AnomalyLikelihood` class. Hence, I added the class name in my test methods to distinguish them from the the test methods that test the low-level functions. Perhaps it would be better to split the tests into two test case classes - something like `AnomalyLikelihoodClassTest` and AnomalyLikelihoodLowLevelTest`? Any suggestions?

scottpurdy commented:
Done with review

scottpurdy commented:
:+1:

scottpurdy commented:
Please update the years in the copyright headers as well

vitaly-krugl commented:
Updated the copyright years

vitaly-krugl commented:
@scottpurdy: I addressed all your feedback via 3ecdd730f219c6a9c04374f254741af4c42edc98. I will merge this PR as soon as the CI build completes and is green.

vitaly-krugl commented:
Merging based on passing Linux builds/tests. Travis is presently having an issue with Mac OS X builds and the ETA is unknown. These changes are not platform-specific, and were also manually tested on Mac OS X.
